auth_test.go

```diff
@@ -0,0 +1,250 @@
+		require.Equal(t, passwordRequest, rw.messages[0])
+		require.True(t, bytes.Contains(rw.messages[1], fatalMarker))
+		require.NoError(t, err)
+	})
+}
+
+func TestAuthenticationMD5_authenticate(t *testing.T) {
+	passwordRequest := msg{
+		'R',
+		0, 0, 0, 12, // length
+		0, 0, 0, 5, // md5 auth type
+	}
+	rw := &mockMD5MessageReadWriter{
+		user: "postgres",
+		pass: []byte("test"),
+		salt: []byte{},
+	}
+	args := map[string]interface{}{
+		"user": "postgres",
+	}
+	pp := &md5ConstantPasswordProvider{password: []byte("test")}
+
+	a := &md5Authenticator{pp}
+
+	t.Run("valid password", func(t *testing.T) {
+		defer rw.Reset()
+		ok, err := a.authenticate(rw, args)
+
+		require.True(t, ok)
+		require.NoError(t, err)
+		require.True(t, bytes.Contains(rw.messages[0], passwordRequest))
+		require.Equal(t, authOKMessage, rw.messages[1])
+	})
+
+	t.Run("invalid password", func(t *testing.T) {
+		defer rw.Read()
+		pp.password = []byte("shtoot")
+		ok, err := a.authenticate(rw, args)
+
+		require.False(t, ok)
+		require.True(t, bytes.Contains(rw.messages[0], passwordRequest))
+		require.True(t, bytes.Contains(rw.messages[1], fatalMarker))
+		require.NoError(t, err)
+	})
+
+	t.Run("invalid message type", func(t *testing.T) {
+		defer rw.Reset()
+		rw := &mockMessageReadWriter{output: []msg{
+			{'q', 0, 0, 0, 5, 1},
+		}}
+		ok, err := a.authenticate(rw, args)
+
+		require.False(t, ok)
+		require.True(t, bytes.Contains(rw.messages[0], passwordRequest))
+		require.True(t, bytes.Contains(rw.messages[1], fatalMarker))
+		require.NoError(t, err)
+	})
+}
+
+func TestHashWithSalt(t *testing.T) {
+	user := "postgres"
+	pass := []byte("test")
+	salt := []byte{196, 53, 49, 235}
+	hash := md5.Sum(append(pass, []byte(user)...))
+
+	// actual hash received from psql using the above variables
+	expectedHash := []byte{
+		109, 100, 53, 97, 97, 51, 102, 56, 98, 56,
+		55, 97, 57, 51, 52, 97, 52, 53, 48, 52,
+		52, 101, 49, 102, 98, 50, 100, 57, 48, 55,
+		48, 99, 98, 56, 48,
+	}
+
+	actualHash := hashWithSalt(hash[:], salt)
+	require.Equal(t, expectedHash, actualHash)
+}
+
+func TestGetRandomSalt(t *testing.T) {
+	var lastSalt []byte
+	for i := 0; i < 100; i++ {
+		salt := getRandomSalt()
+		require.Equal(t, len(salt), 4)
+		require.NotEqual(t, lastSalt, salt)
+		lastSalt = salt
+	}
+}
+
+func TestExtractPassword(t *testing.T) {
+	t.Run("regular password", func(t *testing.T) {
+		passwordMessage := msg{
+			'p',
+			0, 0, 0, 9,
+			42, 42, 42, 42,
+			0,
+		}
+
+		expectedResult := []byte{42, 42, 42, 42}
+		actualResult := extractPassword(passwordMessage)
+		require.Equal(t, expectedResult, actualResult)
+	})
+
+	t.Run("empty password", func(t *testing.T) {
+		passwordMessage := msg{
+			'p',
+			0, 0, 0, 5,
+			0,
+		}
+
+		expectedResult := []byte{}
+		actualResult := extractPassword(passwordMessage)
+		require.Equal(t, expectedResult, actualResult)
+	})
+}
+
+// mockMessageReadWriter implements messageReadWriter and outputs the provided output
```

Comment:
it's [more stub than a mock](https://martinfowler.com/articles/mocksArentStubs.html), but cool tests 🤓